src/core/lib/support/closure_ref.h

```diff
@@ -0,0 +1,248 @@
+const typename ClosureRef<T>::VTable FnClosure1<Scheduler, Env, T, F>::vtable =
+    {FnClosure1<Scheduler, Env, T, F>::Schedule,
+     FnClosure1<Scheduler, Env, T, F>::Run};
+
+template <class Scheduler, class T, void (T::*F)()>
+class MemClosure {
+ public:
+  static const typename ClosureRef<>::VTable vtable;
+
+ private:
+  static void Schedule(void* env) {
+    Scheduler::Schedule([](void* env) { (static_cast<T*>(env)->*F)(); },
+                        static_cast<T*>(env));
+  }
+  static void Run(void* env) {
+    Scheduler::Run([](void* env) { (static_cast<T*>(env)->*F)(); },
+                   static_cast<T*>(env));
+  }
+};
+template <class Scheduler, class T, void (T::*F)()>
+const typename ClosureRef<>::VTable MemClosure<Scheduler, T, F>::vtable = {
+    MemClosure<Scheduler, T, F>::Schedule, MemClosure<Scheduler, T, F>::Run};
+}
+
+template <class Scheduler, void (*F)(), typename Env = std::nullptr_t>
+ClosureRef<> MakeClosure(Env* env = nullptr) {
+  return ClosureRef<>(&impl::FnClosure<Scheduler, Env, F>::vtable, env);
+}
+
+template <class Scheduler, typename T, void (*F)(T),
+          typename Env = std::nullptr_t>
+ClosureRef<T> MakeClosure(Env* env = nullptr) {
+  return ClosureRef<int>(&impl::FnClosure1<Scheduler, Env, T, F>::vtable, env);
+}
+
+template <class Scheduler, typename T, void (T::*F)()>
+ClosureRef<> MakeClosure(T* p) {
+  return ClosureRef<>(&impl::MemClosure<Scheduler, T, F>::vtable, p);
+}
+
+//
+// SCHEDULERS
+//
+
+class AcquiresNoLocks {
+ public:
+  template <class T, class F>
+  static void Schedule(F&& f, T* env) {
+    f(env);
+  }
+  template <class T, class F>
+  static void Run(F&& f, T* env) {
+    f(env);
+  }
+};
+
+template <class F>
+void QueueOnExecCtx(F&& f);
+
+class RunInCurrentThread {
+ public:
+  template <class T, class F>
+  static void Schedule(F&& f, T* env) {
+    QueueOnExecCtx([f, env]() { f(env); });
+  }
+  template <class T, class F>
+  static void Run(F&& f, T* env) {
+    f(env);
+  }
+};
+
+class RunInCombiner {
+ public:
+  template <class T, class F>
+  static void Schedule(F&& f, T* env) {
+    env->combiner()->Schedule([f, env]() { f(env); });
+  }
+  template <class T, class F>
+  static void Run(F&& f, T* env) {
+    env->combiner()->Run([f, env]() { f(env); });
+  }
+};
+
+// Dummy combiner lock impl
+
+class Combiner {
+ public:
+  template <class F>
+  void Schedule(F&& f);
+  template <class F>
+  void Run(F&& f);
+};
+
+//
+// TEST CODE
+//
+
+void PrintLine();
+void PrintInt(int);
+
+class Foo {
+ public:
+  void Callback();
+
+  Combiner* combiner() { return &combiner_; }
+
+ private:
+  Combiner combiner_;
+};
+
+ClosureRef<> Hidden();
+
+void test() {
+  // simple closures around functions, member functions
+  ClosureRef<> print_line = MakeClosure<AcquiresNoLocks, PrintLine>();
+  ClosureRef<int> print_int = MakeClosure<AcquiresNoLocks, int, PrintInt>();
+  Foo foo;
+  ClosureRef<> foo_cb = MakeClosure<AcquiresNoLocks, Foo, &Foo::Callback>(&foo);
```

Comment:
As we've discussed, I expect this to be the common case, and I'd like to see some syntactic sugar to make it less verbose.  Ideally, I'd like to be able to say something like this:

```
ClosureRef<> foo_cb = MakeClosure(AcquiresNoLocks, &foo, &Foo::Callback);
```

Even better would be if the middle arg wasn't needed, but I suspect that would be harder.

Reply:
I think I have a similar but not quite identical issue as Mark. I'm wondering why the callable should be a template parameter as opposed to a function parameter at the API level. (I'm fine with whatever is needed at the implementation level.)

Reply:
I made one tweak... to move the scheduler out of the argument list here, so now you write:

```c++
SchedulerType::MakeClosure<Foo, &Foo::Callback>();
```

Too address @vjpai's comment below also: it's needed to be in the template parameter list to save needing to store an additional word in ClosureRef (instead adding a little bit to the library size for each closure - which is a good tradeoff for us I believe).

Imagine this more like a code-generator that runs at compile time and generates specialized code to call the passed in function within the selected scheduler.

Reply:
`ScheduleType::MakeClosure` is very nice.

I'm pretty template friendly in general, but using a callable as a template parameter isn't yet common for our repo, so it jars a little. I'm ok with the performance tradeoff, though.